tests/unit/ConstantCastingSpec.hs

```diff
@@ -0,0 +1,99 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module ConstantCastingSpec (spec) where
+
+import Control.Monad
+import Test.Hspec
+import Test.Hspec.QuickCheck
+import Test.QuickCheck
+
+import Builder
+import Core
+import Lower
+import Name
+import Optimize
+import TopLevel
+import Types.Core
+import Types.Imp
+import Types.Primitives
+import Util
+
+castOp :: ScalarBaseType -> LitVal -> PrimOp (SAtom VoidS)
+castOp ty x = MiscOp $ CastOp (BaseTy (Scalar ty)) (Con (Lit x))
+
+exprToBlock :: EnvReader m => Expr r n -> m n (Block r n)
+exprToBlock expr = do
+  liftBuilder $ buildBlock $ do
+    v <- emit $ sink expr
+    return $ Var v
+
+evalBlock :: (Topper m, Mut n) => SBlock n -> m n (SAtom n)
+evalBlock block = lowerFullySequential block >>= evalLLVM
+
+evalClosedExpr :: SExpr VoidS -> IO LitVal
+evalClosedExpr expr = do
+  let cfg = EvalConfig LLVM [LibBuiltinPath] Nothing Nothing Nothing NoOptimize
+  env <- initTopState
+  fst <$> runTopperM cfg env do
+    block <- exprToBlock $ unsafeCoerceE expr
+    (Var name) <- evalBlock block
+    (AtomNameBinding (TopDataBound (RepVal _ (Leaf (ILit ans))))) <- lookupEnv name
+    return ans
+
+instance Arbitrary LitVal where
+  arbitrary = oneof
+    [ Int64Lit   <$> arbitrary
+    , Int32Lit   <$> arbitrary
+    , Word8Lit   <$> arbitrary
+    , Word32Lit  <$> arbitrary
+    , Word64Lit  <$> arbitrary
+    , Float64Lit <$> arbitrary
+    , Float32Lit <$> arbitrary
+    ]
+
+spec :: Spec
+spec = do
+  describe "constant-folding casts" do
+    let constant_folding_and_runtime_casts_agree ty = 
+          \(x::LitVal) -> case foldCast ty x of
+              Nothing -> return ()
+              Just folded -> do
+                let op = castOp ty x
+                evaled <- evalClosedExpr $ PrimOp op
+                -- The failure message will list `evaled` as "expected" and `folded` as "got"
+                folded `shouldBe` evaled
+    forM_ [Int64Type, Int32Type, Word8Type, Word32Type, Word64Type, Float64Type, Float32Type] \ty ->
+      -- TODO: We'd really like to run 10,000 or 1,000,000 examples here, but
+      -- status quo is that each one runs through the LLVM compile-and-run
+      -- pipeline separately, and is thus incredibly slow.  Taking 50 as a
+      -- compromise between test suite speed and test coverage.
+      -- I ran this offline with 10,000 before checking in, and it passed.
```

Comment:
To be fair, we do have everything we need to compile the program that does the rounding once and evaluate it repeatedly on different inputs. But we can leave it as is for now.

Reply:
Ooo, how?

Reply:
If you look at the body of [`evalLLVM`](https://github.com/google-research/dex-lang/blob/main/src/lib/TopLevel.hs#L773), the majority of it leads to generating `nativeFun`, which is then only called once and discarded. What you could do, is to compile a `Float -> Float` function and then use `callNativeFun` repeatedly to evaluate it over a range of inputs.